File: ballast-core/dotnet/src/messaging/EventBase.cs
using System;
using Ballast.Core.Utilities;

namespace Ballast.Core.Messaging
{
    public abstract class EventBase : IEvent 
    { 

        public abstract string Id { get; }

        public virtual string EventDateIsoString { get; protected set; }

        public EventBase(string eventDateIsoString = null) 
        {
            EventDateIsoString = eventDateIsoString ?? EventBase.GetDateIsoString();
        }

        public static string GetDateIsoString() => 
            DateTime.UtcNow.ToIsoString();
        
    }
}

File: ballast-core/npm/src/messaging/event-base.ts
import { IEvent } from "./event";
import { getUtcNow } from "../utility/date-helpers";

export abstract class EventBase implements IEvent { 
    public abstract id: string;
    public eventDateIsoString: string;
    public constructor(dateIsoString?: string) {
        this.eventDateIsoString = dateIsoString || EventBase.getDateIsoString();
    }
    public static getDateIsoString() {
        return getUtcNow().toISOString();
    }
}